### PR TITLE
Deprecate secret token

### PR DESCRIPTION
**Pull request by claudiob**
This is another take on #30138 where secret_token is more strongly deprecated rather than removed, following comments by @dhh and @rafaelfranca on Campfire.

I have created three commits to deal separately with Action Pack, Guides, and Railties.

I'm not 100% sure about this PR… let me know if that's what you meant by "stronger deprecation in 5.2". Thanks!

### CODE REVIEW

File: railties/lib/rails/application.rb [hunk 1]
```diff
@@ -511,6 +515,11 @@ def validate_secret_key_config! #:nodoc:
 
         if secrets.secret_token.blank?
           raise "Missing `secret_key_base` for '#{Rails.env}' environment, set this value in `config/secrets.yml`"
+        else
+          ActiveSupport::Deprecation.warn(<<-MSG.squish)
+           secrets.secret_token is deprecated and will be removed in Rails 6.0.
+           Please use secrets.secret_key_base instead.
+          MSG
```
Comment (kaspth) [inline railties/lib/rails/application.rb]: Let's unify these 2 deprecation warnings into here and wrap it in `secrets.secret_token.present?`: https://github.com/rails/rails/blob/db844d9fb30cd34411fcdcc6780f82aaa5471e81/railties/lib/rails/application.rb#L395

File: actionpack/lib/action_dispatch/middleware/cookies.rb [hunk 2]
```diff
@@ -50,7 +50,14 @@ def authenticated_encrypted_cookie_salt
...
     def secret_token
-      get_header Cookies::SECRET_TOKEN
+      secret_token = get_header Cookies::SECRET_TOKEN
+      if secret_token
+        ActiveSupport::Deprecation.warn(<<-MSG.squish)
+         cookies.secret_token is deprecated and will be removed in Rails 6.0.
+         Please use cookies.secret_key_base instead.
+        MSG
+      end
+      secret_token
```
Comment (kaspth) [inline actionpack/lib/action_dispatch/middleware/cookies.rb]: This is an internal method (note the `# :stopdoc:` above), so we don't need this deprecation warning or the Action Pack changelog entry.

File: railties/lib/rails/application.rb [hunk 3]
```diff
@@ -511,6 +515,11 @@ def validate_secret_key_config! #:nodoc:
 
         if secrets.secret_token.blank?
           raise "Missing `secret_key_base` for '#{Rails.env}' environment, set this value in `config/secrets.yml`"
+        else
+          ActiveSupport::Deprecation.warn(<<-MSG.squish)
+           secrets.secret_token is deprecated and will be removed in Rails 6.0.
+           Please use secrets.secret_key_base instead.
```
Comment (kaspth) [inline railties/lib/rails/application.rb]: Let's say: "Instead, assign a `secret_key_base` through either credentials, encrypted secrets or in `config/secrets.yml`."

File: actionpack/test/dispatch/cookies_test.rb [hunk 4]
```diff
@@ -911,7 +921,7 @@ def test_legacy_signed_cookie_is_treated_as_nil_by_encrypted_cookie_jar_if_tampe
     @request.env["action_dispatch.secret_key_base"] = "c3b95688f35581fad38df788add315ff"
 
     @request.headers["Cookie"] = "foo=baz"
-    get :get_encrypted_cookie
+    ActiveSupport::Deprecation.silence { get :get_encrypted_cookie }
```
Comment (kaspth) [inline actionpack/test/dispatch/cookies_test.rb]: Don't think we need these silences with the Action Pack deprecation removed.

**Review by kaspth — COMMENTED** [review]
We should probably also add a test in `test/application/configuration_test.rb` that assigns a `secret_token` ala:

```ruby
app_file "config/secrets.yml", <<-RUBY
  production:
    secret_key_base: blargh
    secret_token: deprecate-me
RUBY

assert_deprecated(/secret_token is deprecated/) do
  app "production"
end
```

**Comment by claudiob** [comment]
@kaspth Thanks for your feedback. 
Given that #30067 has now been merged, I will review these changes and ping you back when I'm done.

**Comment by kaspth** [comment]
Hey, had time to look into this?

**Comment by claudiob** [comment]
@kaspth I was on vacation! I'm a little lost now between secrets, credentials… 
I'm not even sure we need this PR anymore since it's deprecating something that was already deprecated. What do you think? Should I just close this PR?

**Comment by kaspth** [comment]
No problem. `secrets` as a whole hasn't been deprecated yet, I'm still mulling that over. I handled the deprecation in fbcc4bfe9a, partially based on this so you got the credit as well. Thanks!

**Pull request closed without merging**